scripts/install_xcpng.py

```diff
@@ -0,0 +1,219 @@
+    fb = open(bootfile, 'w')
+    x = f"""answerfile=custom
+installer={installer}
+is_default=1
+"""
+    fb.write(x)
+    fb.close()
+
+def generate_answerfile(tmp_directory, installer, hostname_or_ip, type, hdd):
+    pxe = pxe_address()
+    password = host_data(hostname_or_ip)['password']
+    tempstream = os.popen(f"mkpasswd -m SHA-512 '{password}'")
+    encrypted_password = tempstream.read()
+    answerfile = f'{tmp_directory}/answerfile.xml'
+    fa = open(answerfile, 'w')
+    if type == 'install':
+        y = f"""<?xml version="1.0"?>
+    <installation>
+        <keymap>fr</keymap>
+        <primary-disk>{hdd}</primary-disk>
+        <guest-disk>{hdd}</guest-disk>
+        <root-password type="hash">{encrypted_password}</root-password>
+        <source type="url">{installer}</source>
+        <admin-interface name="eth0" proto="dhcp" />
+        <timezone>Europe/Paris</timezone>
+        <script stage="filesystem-populated" type="url">
+        http://{pxe}/configs/presets/scripts/filesystem-populated.py
+        </script>
+    </installation>
+    """
+    elif type == 'upgrade':
+        y = f"""<?xml version="1.0"?>
+    <installation mode="upgrade">
+    <existing-installation>sda</existing-installation>
+    <source type="url">{installer}</source>
+    <script stage="filesystem-populated" type="url">
+        http://{pxe}/configs/presets/scripts/filesystem-populated.py
+        </script>
+    </installation>
+    """
+    fa.write(y)
+    fa.close()
+
+def copy_files_to_pxe(mac_address, tmp_local_path):
+    pxe = pxe_address()
+    distant_directory = f'/pxe/configs/custom/{mac_address}/'
+    clean_files_on_pxe(mac_address)
+    ssh(pxe, ['mkdir', '-p', distant_directory])
+    scp(pxe, f'{tmp_local_path}/boot.conf', distant_directory)
+    scp(pxe, f'{tmp_local_path}/answerfile.xml', distant_directory)
+
+def clean_files_on_pxe(mac_address):
+    pxe = pxe_address()
+    if mac_address:
+        distant_directory = f'/pxe/configs/custom/{mac_address}/'
+        ssh(pxe, ['rm', '-rf', distant_directory])
+    else:
+        raise Exception('ERROR: mac address is null or empty!')
+
+def clean_bootconf_on_pxe(mac_address):
+    pxe = pxe_address()
+    if mac_address:
+        distant_file = f'/pxe/configs/custom/{mac_address}/boot.conf'
+        try:
+            ssh(pxe, ['rm', '-rf', distant_file])
+        except SSHCommandFailed as e:
+            raise Exception('ERROR: problem to clean the boot.conf file.' + e)
+    else:
+        raise Exception('ERROR: mac address is null or empty!')
+
+def get_candidate_ips(mac_address):
+    pxe = pxe_address()
+    ip_address = ssh(
+        pxe,
+        ['arp', '-n', '|', 'grep', mac_address, '|', 'awk', '\'{ print $1 }\'']
+    )
+    candidate_ips = ip_address.splitlines()
+    return candidate_ips
+
+def is_ip_active(ip):
+    ret = os.system(f"ping -c 3 -W 10 {ip} > /dev/null 2>&1")
+    if ret != 0:
+        return False
+    else:
+        return True
```

Comment:
```suggestion
    return ret == 0
```

Reply:
i don't understand that one. why adding another return when we already have what we need ?
why returning something we don't use after ?

Reply:
```python
return not os.system(f"ping -c 3 -W 10 {ip} > /dev/null 2>&1")
```

Reply:
but i don't want to return 0.
so i don't understand the logic.

Reply:
```python
return ret == 0 # It returns a boolean.
not 0 # Same.
```

Reply:
but ret isn't a boolean, si if i don't test it, how can i know if it's true or false ? ret is an integer.

Reply:
`ret == 0`: It's a boolean expression, the result of the call to the `==` operator is always a boolean.
`not 0`: Same, it's a boolean expression, `not` on an integer returns True if the value is 0. And False for the other cases.

Reply:
Test in a python interpreter. ;)

Reply:
Said in a different way: you'll not be returning `ret`, you'll return whatever `ret == 0` evaluates to, that is either `True` or `False`.

Reply:
Illustration:
```
>>> def is_one(value):
...     return value == 1
... 
>>> is_one(2)
False
>>> is_one(1)
True
```